# Akash patil
aakashgolu1008@gmail.com

## Support Vector Machines (SVMs)

Prepare a classification model using SVM for salary data 

Data Description:

age -- age of a person <br>
workclass	-- A work class is a grouping of work <br>
education	-- Education of an individuals	<br>
maritalstatus -- Marital status of an individulas <br>
occupation	 -- occupation of an individuals <br>
relationship -- 	
race --  Race of an Individual<br>
sex --  Gender of an Individual <br>
capitalgain --  profit received from the sale of an investment	<br>
capitalloss	-- A decrease in the value of a capital asset<br>
hoursperweek -- number of hours work per week	<br>
native -- Native of an individual<br>
Salary -- salary of an individual<br>


In [1]:
import pandas as pd
import numpy as np

In [2]:
sd_train=pd.read_csv("SalaryData_Train.csv")
sd_train.sample(n=10)

age          workclass      education  educationno  \
12105   25            Private        1st-4th            2   
22365   66            Private        HS-grad            9   
13710   48            Private        HS-grad            9   
1686    56            Private        HS-grad            9   
9439    39   Self-emp-not-inc   Some-college           10   
16568   31          State-gov           12th            8   
10924   26            Private        Masters           14   
27284   23            Private   Some-college           10   
8917    47          Local-gov   Some-college           10   
20578   54          State-gov      Bachelors           13   

             maritalstatus          occupation    relationship    race  \
12105        Never-married     Priv-house-serv   Not-in-family   White   
22365   Married-civ-spouse   Handlers-cleaners         Husband   White   
13710             Divorced   Machine-op-inspct   Not-in-family   White   
1686         Never-married        Craft-repair   Not-in-family   White   
9439         Never-married        Craft-repair   Not-in-family   White   
16568   Married-civ-spouse        Tech-support            Wife   White   
10924        Never-married     Exec-managerial       Own-child   White   
27284        Never-married   Handlers-cleaners       Own-child   Black   
8917              Divorced    Transport-moving       Unmarried   White   
20578   Married-civ-spouse     Exec-managerial         Husband   White   

           sex  capitalgain  capitalloss  hoursperweek          native  Salary  
12105   Female            0            0            40       Guatemala   <=50K  
22365     Male            0            0            40   United-States   <=50K  
13710     Male            0            0            40   United-States   <=50K  
1686      Male            0            0            40   United-States   <=50K  
9439      Male            0            0            25   United-States   <=50K  
16568   Female            0            0            40   United-States    >50K  
10924     Male            0            0            38   United-States   <=50K  
27284   Female            0            0            40   United-States   <=50K  
8917      Male            0            0            40   United-States   <=50K  
20578     Male            0            0            40     Puerto-Rico    >50K

In [3]:
sd_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30161 entries, 0 to 30160
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   age            30161 non-null  int64 
 1   workclass      30161 non-null  object
 2   education      30161 non-null  object
 3   educationno    30161 non-null  int64 
 4   maritalstatus  30161 non-null  object
 5   occupation     30161 non-null  object
 6   relationship   30161 non-null  object
 7   race           30161 non-null  object
 8   sex            30161 non-null  object
 9   capitalgain    30161 non-null  int64 
 10  capitalloss    30161 non-null  int64 
 11  hoursperweek   30161 non-null  int64 
 12  native         30161 non-null  object
 13  Salary         30161 non-null  object
dtypes: int64(5), object(9)
memory usage: 3.2+ MB


In [4]:
sd_train.isnull().sum()

age              0
workclass        0
education        0
educationno      0
maritalstatus    0
occupation       0
relationship     0
race             0
sex              0
capitalgain      0
capitalloss      0
hoursperweek     0
native           0
Salary           0
dtype: int64

In [5]:
col=sd_train.select_dtypes(include=object).columns
col

Index(['workclass', 'education', 'maritalstatus', 'occupation', 'relationship',
       'race', 'sex', 'native', 'Salary'],
      dtype='object')

In [6]:

sd_train['workclass']=sd_train['workclass'].astype('category')
sd_train['education']=sd_train['education'].astype('category')
sd_train['maritalstatus']=sd_train['maritalstatus'].astype('category')
sd_train['occupation']=sd_train['occupation'].astype('category')
sd_train['relationship']=sd_train['relationship'].astype('category')
sd_train['race']=sd_train['race'].astype('category')
sd_train['native']=sd_train['native'].astype('category')
sd_train['sex']=sd_train['sex'].astype('category')

In [7]:
sd_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30161 entries, 0 to 30160
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype   
---  ------         --------------  -----   
 0   age            30161 non-null  int64   
 1   workclass      30161 non-null  category
 2   education      30161 non-null  category
 3   educationno    30161 non-null  int64   
 4   maritalstatus  30161 non-null  category
 5   occupation     30161 non-null  category
 6   relationship   30161 non-null  category
 7   race           30161 non-null  category
 8   sex            30161 non-null  category
 9   capitalgain    30161 non-null  int64   
 10  capitalloss    30161 non-null  int64   
 11  hoursperweek   30161 non-null  int64   
 12  native         30161 non-null  category
 13  Salary         30161 non-null  object  
dtypes: category(8), int64(5), object(1)
memory usage: 1.6+ MB


In [14]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder().fit_transform

In [17]:
sd_train['Salary']=le(sd_train["Salary"])

In [19]:
sd_train.Salary.head()

0    0
1    0
2    0
3    0
4    0
Name: Salary, dtype: int32

In [22]:
col=sd_train.select_dtypes(include='category').columns
col

Index(['workclass', 'education', 'maritalstatus', 'occupation', 'relationship',
       'race', 'sex', 'native'],
      dtype='object')

In [26]:
for i in col:
    sd_train[i]=le(sd_train[i])
sd_train.sample(n=10)

age  workclass  education  educationno  maritalstatus  occupation  \
24828   62          2          4            3              2           4   
28787   55          1         12           14              0           9   
11495   40          2         15           10              2           2   
14868   39          2         11            9              0           0   
22003   19          2         11            9              4          11   
18433   46          2         11            9              0           8   
19624   24          1         15           10              4          10   
8498    35          2          9           13              2           7   
24147   60          2         12           14              2          13   
6386    24          2          9           13              4           9   

       relationship  race  sex  capitalgain  capitalloss  hoursperweek  \
24828             0     1    1            0            0            40   
28787             4     4    0            0            0            45   
11495             0     4    1            0            0            40   
14868             4     4    0            0            0            40   
22003             4     2    0            0            0            30   
18433             1     4    0            0            0            25   
19624             4     3    1         1151            0            40   
8498              5     4    0            0            0            24   
24147             0     4    1            0            0            10   
6386              3     1    1            0            0            40   

       native  Salary  
24828      28       0  
28787      37       0  
11495      37       0  
14868      37       0  
22003      37       0  
18433      37       0  
19624      37       0  
8498       29       0  
24147      37       0  
6386       37       0

In [28]:
# Splitting the data into x and y as input and output
x=sd_train.iloc[:,:13]
y=sd_train.iloc[:,13]

In [29]:
x.sample(n=5)

age  workclass  education  educationno  maritalstatus  occupation  \
24328   39          4         14           15              2           9   
10909   40          1         15           10              2           0   
27542   30          2         11            9              4           8   
8046    40          2         11            9              2           3   
931     26          2         15           10              2          11   

       relationship  race  sex  capitalgain  capitalloss  hoursperweek  native  
24328             0     4    1            0            0            45      37  
10909             5     4    0            0            0            40      37  
27542             3     4    0            0            0            40      12  
8046              0     4    1            0         1672            45      37  
931               0     4    1            0            0            50      10

In [30]:
y.sample(n=5)

23858    0
24325    0
3485     0
7438     0
15593    0
Name: Salary, dtype: int32

In [31]:
sd_train.Salary.value_counts()

0    22653
1     7508
Name: Salary, dtype: int64

In [86]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(x,y,test_size=0.30,random_state=1)
X_train

age  workclass  education  educationno  maritalstatus  occupation  \
10783   36          2         12           14              0           9   
9431    23          2         11            9              4           6   
14839   63          4         14           15              2           9   
4730    24          4          1            7              4           7   
2900    57          2          8           11              3           3   
...    ...        ...        ...          ...            ...         ...   
905     60          4          6            5              2           3   
5192    20          2         11            9              4           2   
12172   40          2          7           12              5           6   
235     62          0         15           10              0          12   
13349   31          2         11            9              2          13   

       relationship  race  sex  capitalgain  capitalloss  hoursperweek  native  
10783             1     4    0            0            0            40      37  
9431              2     2    1            0            0            40      37  
14839             0     4    1            0            0            35      37  
4730              4     0    0            0            0            40      37  
2900              1     4    0            0            0            47      37  
...             ...   ...  ...          ...          ...           ...     ...  
905               0     4    1            0         1887            67      37  
5192              3     4    1            0            0            40      37  
12172             1     4    1            0            0            40      26  
235               1     2    1         4650            0            40      37  
13349             0     4    1            0            0            50      37  

[10542 rows x 13 columns]

In [87]:
y_test

3600     0
5264     1
11606    1
9784     0
12997    0
        ..
8365     0
6496     0
9017     0
11852    0
12164    0
Name: Salary, Length: 4518, dtype: int32

In [88]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
x_train=scaler.fit_transform(X_train)
x_test=scaler.transform(X_test)

In [89]:
from sklearn.svm import SVC
clf=SVC()
clf.fit(x_train,y_train)
pred=clf.predict(x_test)

In [90]:
pred

array([0, 1, 0, ..., 0, 0, 0])

In [91]:
from sklearn.metrics import accuracy_score,classification_report,confusion_matrix
print('accuracy_score', accuracy_score(pred,y_test))
print('classification_report\n', classification_report(pred,y_test))
confusion_matrix(pred,y_test)

accuracy_score 0.83444001770695
classification_report
               precision    recall  f1-score   support

           0       0.92      0.87      0.89      3656
           1       0.55      0.70      0.62       862

    accuracy                           0.83      4518
   macro avg       0.74      0.78      0.76      4518
weighted avg       0.85      0.83      0.84      4518



array([[3166,  490],
       [ 258,  604]], dtype=int64)

### test data set

In [44]:
sd_test=pd.read_csv("SalaryData_Test.csv")
sd_test.sample(n=5)

age          workclass   education  educationno        maritalstatus  \
11572   21          Local-gov        11th            7        Never-married   
3556    65            Private        11th            7             Divorced   
14328   22            Private     HS-grad            9        Never-married   
4476    52   Self-emp-not-inc     HS-grad            9   Married-civ-spouse   
7086    22          State-gov   Bachelors           13        Never-married   

               occupation    relationship    race      sex  capitalgain  \
11572       Other-service       Own-child   Black     Male            0   
3556    Handlers-cleaners   Not-in-family   White     Male            0   
14328        Adm-clerical       Own-child   White   Female            0   
4476         Craft-repair         Husband   White     Male            0   
7086       Prof-specialty   Not-in-family   White     Male            0   

       capitalloss  hoursperweek          native  Salary  
11572            0            25   United-States   <=50K  
3556             0            24   United-States   <=50K  
14328            0            40   United-States   <=50K  
4476             0            50           India    >50K  
7086             0            15   United-States   <=50K

In [46]:
sd_test.Salary.nunique()

2

In [47]:
sd_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15060 entries, 0 to 15059
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   age            15060 non-null  int64 
 1   workclass      15060 non-null  object
 2   education      15060 non-null  object
 3   educationno    15060 non-null  int64 
 4   maritalstatus  15060 non-null  object
 5   occupation     15060 non-null  object
 6   relationship   15060 non-null  object
 7   race           15060 non-null  object
 8   sex            15060 non-null  object
 9   capitalgain    15060 non-null  int64 
 10  capitalloss    15060 non-null  int64 
 11  hoursperweek   15060 non-null  int64 
 12  native         15060 non-null  object
 13  Salary         15060 non-null  object
dtypes: int64(5), object(9)
memory usage: 1.6+ MB


In [57]:
sd_test.select_dtypes(include=object).columns

Index(['workclass', 'education', 'maritalstatus', 'occupation', 'relationship',
       'race', 'sex', 'native', 'Salary'],
      dtype='object')

In [58]:
sd_test['workclass']=sd_test['workclass'].astype('category')
sd_test['education']=sd_test['education'].astype('category')
sd_test['maritalstatus']=sd_test['maritalstatus'].astype('category')
sd_test['occupation']=sd_test['occupation'].astype('category')
sd_test['relationship']=sd_test['relationship'].astype('category')
sd_test['race']=sd_test['race'].astype('category')
sd_test['native']=sd_test['native'].astype('category')
sd_test['sex']=sd_test['sex'].astype('category')

In [59]:
sd_test['Salary']=le(sd_test['Salary'])

In [62]:
Col=sd_test.select_dtypes(include='category').columns
Col

Index(['workclass', 'education', 'maritalstatus', 'occupation', 'relationship',
       'race', 'sex', 'native'],
      dtype='object')

In [63]:
for i in Col:
    sd_test[i]=le(sd_test[i])
sd_test

age  workclass  education  educationno  maritalstatus  occupation  \
0       25          2          1            7              4           6   
1       38          2         11            9              2           4   
2       28          1          7           12              2          10   
3       44          2         15           10              2           6   
4       34          2          0            6              4           7   
...    ...        ...        ...          ...            ...         ...   
15055   33          2          9           13              4           9   
15056   39          2          9           13              0           9   
15057   38          2          9           13              2           9   
15058   44          2          9           13              0           0   
15059   35          3          9           13              2           3   

       relationship  race  sex  capitalgain  capitalloss  hoursperweek  \
0                 3     2    1            0            0            40   
1                 0     4    1            0            0            50   
2                 0     4    1            0            0            40   
3                 0     2    1         7688            0            40   
4                 1     4    1            0            0            30   
...             ...   ...  ...          ...          ...           ...   
15055             3     4    1            0            0            40   
15056             1     4    0            0            0            36   
15057             0     4    1            0            0            50   
15058             3     1    1         5455            0            40   
15059             0     4    1            0            0            60   

       native  Salary  
0          37       0  
1          37       0  
2          37       1  
3          37       1  
4          37       0  
...       ...     ...  
15055      37       0  
15056      37       0  
15057      37       0  
15058      37       0  
15059      37       1  

[15060 rows x 14 columns]

In [64]:
x=sd_test.iloc[:,:13]
y=sd_test.iloc[:,13]

In [65]:
x.sample(n=5)

age  workclass  education  educationno  maritalstatus  occupation  \
9362    35          1         15           10              0          12   
11633   29          2         11            9              0           6   
13098   35          2         15           10              0           9   
970     40          2          0            6              0           2   
14894   42          1         12           14              4           9   

       relationship  race  sex  capitalgain  capitalloss  hoursperweek  native  
9362              4     4    0            0            0            40      37  
11633             1     4    1            0            0            50      37  
13098             1     4    0            0            0            40      37  
970               3     4    1            0            0            40      37  
14894             1     4    0            0            0            45      37

In [93]:
test=scaler.fit_transform(x)
test

array([[-1.02900513, -0.222347  , -2.42343884, ..., -0.2191694 ,
        -0.07888904,  0.26058067],
       [-0.05742253, -0.222347  ,  0.19017232, ..., -0.2191694 ,
         0.75013125,  0.26058067],
       [-0.80479376, -1.25712632, -0.85527215, ..., -0.2191694 ,
        -0.07888904,  0.26058067],
       ...,
       [-0.05742253, -0.222347  , -0.33254991, ..., -0.2191694 ,
         0.75013125,  0.26058067],
       [ 0.39100021, -0.222347  , -0.33254991, ..., -0.2191694 ,
        -0.07888904,  0.26058067],
       [-0.2816339 ,  0.81243231, -0.33254991, ..., -0.2191694 ,
         1.57915155,  0.26058067]])

In [96]:
pred_t=clf.predict(test)

In [97]:
accuracy_score(pred_t,y)

0.849136786188579

### Salary train dataset accuracy is 0.83

### Salary test dataset accuracy is 0.84